modules/rostests/kmtests/ntos_ob/ObInsert.c

```diff
@@ -0,0 +1,651 @@
+
+    /* Initialize the object name */
+    RtlInitUnicodeString(&ObjectName,
+                         L"\\ObInsertTest\\DummyObject_Success");
+
+    /* Initialize object attributes with the name */
+    InitializeObjectAttributes(&DummyObjectAttributes,
+                               &ObjectName,
+                               OBJ_CASE_INSENSITIVE,
+                               NULL,
+                               NULL);
+
+    /* Create the object */
+    Status = ObCreateObject(KernelMode,
+                            DummyType,
+                            &DummyObjectAttributes,
+                            KernelMode,
+                            NULL,
+                            sizeof(DUMMY_TYPE),
+                            0,
+                            0,
+                            (PVOID *)&Object);
+
+    ok_eq_hex(Status, STATUS_SUCCESS);
+
+    /* Initialize the object data */
+    RtlZeroMemory(Object, sizeof(DUMMY_TYPE));
+    Object->Data = 123;
+
+    ObjectHeader = OBJECT_TO_OBJECT_HEADER(Object);
+
+    ok_eq_ulong(ObjectHeader->PointerCount, 1LU);
+    ok_eq_ulong(ObjectHeader->HandleCount, 0LU);
+
+    /* Insert the object */
+    Status = ObInsertObject(Object,
+                            NULL,
+                            STANDARD_RIGHTS_ALL,
+                            0,
+                            NULL,
+                            &Handle);
+
+    ok_eq_hex(Status, STATUS_SUCCESS);
+    ok(Object != NULL, "Object is NULL\n");
+    ok(Handle != NULL, "Handle is NULL\n");
+
+    ok_eq_ulong(ObjectHeader->PointerCount, 2LU);
+    ok_eq_ulong(ObjectHeader->HandleCount, 1LU);
+
+    ObDereferenceObject(Object);
+
+    Status = ZwClose(Handle);
+    ok_eq_hex(Status, STATUS_SUCCESS);
+}
+
+/*!
+ * @brief This function tests STATUS_OBJECT_PATH_NOT_FOUND.
+ */
+static
+VOID
+ObInsert_PathNotFound()
+{
+    NTSTATUS Status;
+    UNICODE_STRING ObjectName;
+    PDUMMY_TYPE Object;
+    HANDLE Handle;
+    POBJECT_HEADER ObjectHeader;
+
+    /* Initialize the object name. To simulate STATUS_OBJECT_PATH_NOT_FOUND, provide
+       an invalid non-last path component ("ObInsertTestt") */
+    RtlInitUnicodeString(&ObjectName,
+                         L"\\ObInsertTestt\\DummyObject_PathNotFound");
+
+    /* Initialize object attributes with the name */
+    InitializeObjectAttributes(&DummyObjectAttributes,
+                               &ObjectName,
+                               OBJ_CASE_INSENSITIVE,
+                               NULL,
+                               NULL);
+
+    /* Create the object */
+    Status = ObCreateObject(KernelMode,
+                            DummyType,
+                            &DummyObjectAttributes,
+                            KernelMode,
+                            NULL,
+                            sizeof(DUMMY_TYPE),
+                            0,
+                            0,
+                            (PVOID *)&Object);
+
+    ok_eq_hex(Status, STATUS_SUCCESS);
```

Comment:
The creation would succeed, even though the path given via the `DummyObjectAttributes` contains a non-existing path component?? weird.

Reply:
`ObCreateObject` prepares the object header for `ObInsertObject` as needed depending on how the attributes were initialized. However, the checks themselves are mainly in the function that creates the object instance and adds it to the table, namely `ObInsertObject`.